### PR TITLE
Add tgz package and tarball tools

### PR DESCRIPTION
**Pull request by wangkuiyi**
I tested the tools with the PNG-format MNIST datset https://github.com/myleott/mnist_png

### CODE REVIEW

File: tool/tgz/reader.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	if e != nil {
+		return nil, e
+	}
+
+	r.f = f
+	return r, nil
+}
+
+// NewReader wraps an io.Reader into a Reader.
+func NewReader(r io.Reader) (*Reader, error) {
```
Comment (typhoonzero) [inline tool/tgz/reader.go:34]: Is `NewReader` only used in package `tgz`? If so we should not expose it.

Comment (wangkuiyi) [reply tool/tgz/reader.go:34]: I am working on make `ImageNetLoader` call `tgz.NewReader`. It should be used with the MNIST loader (the loading .tgz version) too.

File: tool/tgz/list.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
+		case tar.TypeReg:
+			l = append(l, hdr)
+			io.Copy(ioutil.Discard, r)
+		}
+	}
+	return l, nil
```
Comment (Yancey1989) [inline tool/tgz/list.go:34]: This line is unreachable code, maybe we can remove it.

File: tool/tgz/list.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package tgz
+
+import (
+	"archive/tar"
+	"io"
+	"io/ioutil"
+)
+
+// ListFile list contents in a .tar.gz file.
```
Comment (Yancey1989) [inline tool/tgz/list.go:9]: > ListFile list contents

=>

> ListFile lists regular files header.

It seems this function only lists the regular file header.

File: tool/tgz/testing.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+		t.Fatalf("Cannot create writer: %v", e)
+	}
+
+	imgFns := []string{
+		"mnist/training/0/first.png",
+		"mnist/training/0/second.png",
+		"mnist/training/1/first.png",
+		"mnist/training/1/second.png",
+		"mnist/testing/0/first.png",
+	}
```
Comment (Yancey1989) [inline tool/tgz/testing.go:29]: Maybe we can add some folders that would be more authentically.

``` golang
imgFolders := []string {
    "mnist",
     "mnist/training/",
     "mnist/testing",
     "mnist/training/0",
     "mnist/training/1",
      "mnist/testing/0"
}
```

**Review by Yancey1989 — APPROVED** [review]
LGTM with a tiny comment, please feel free to merge it and update in the next PR.

File: tool/tgz/list.go (hunk 3 above)
Comment (wangkuiyi) [reply tool/tgz/list.go]: Done

File: tool/tgz/list.go (hunk 2 above)
Comment (wangkuiyi) [reply tool/tgz/list.go:38]: Changed the previous `return l, nil ` into `break`.

File: tool/tgz/testing.go (hunk 4 above)
Comment (wangkuiyi) [reply tool/tgz/testing.go:36]: Done

**Review by Yancey1989 — APPROVED** [review]
LGTM!

**Pull request merged**